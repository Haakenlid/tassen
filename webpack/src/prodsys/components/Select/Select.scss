@import 'styles/globals.scss';
$blue: #2684ff;
$selected-color: $blue;
$active-color: lighten($blue, 15%);
$focused-color: lighten($blue, 25%);

.PhotoOption {
  font-size: 0.9rem;
  $size: 6rem;
  $row: $size / 4;
  align-items: center;
  gap: 0 0.5rem;
  min-height: $size;
  display: grid;
  /* prettier-ignore */
  grid:
    "thumb filename filename" #{$row}
    "thumb artist category" #{$row}
    "thumb created size" #{$row}
    "description description description" minmax(#{$row}, auto)
  / minmax(25%, #{$size/2}) 2fr 1fr;

  @each $field in thumb, filename, category, artist, created, size {
    .#{$field} {
      grid-area: #{$field};
      text-overflow: ellipsis;
      white-space: nowrap;
      overflow: hidden;
      border-bottom: 1px solid #eee;
      min-width: 1em;
    }
  }
  .description {
    grid-area: description;
  }
  .filename {
    font-size: 1rem;
    font-weight: bold;
  }
  .popup {
    position: absolute;
    height: auto;
    display: none;
    pointer-events: none;
  }
  .thumb {
    height: 100%;
    border: none;
    &:hover .popup {
      display: block;
    }
  }
}

.ContributorOption {
  $size: 1.5rem;
  gap: 0 0.5em;
  display: grid;
  grid: 'thumb name' auto 'thumb title' auto / auto 1fr;
  align-items: center;
  .thumb {
    grid-area: thumb;
    height: 2 * $size;
    width: auto;
    object-fit: cover;
    border: 1px solid #ddd;
    border-radius: 100%;
  }
  .name {
    font-weight: bold;
    border-bottom: 1px solid #eee;
  }
}

.ReactSelect {
  position: relative;
}
.react-select {
  &__indicators {
    line-height: 0;
    display: flex;
    height: 1rem;
    align-items: center;
  }
  &__indicator {
    width: 0.9rem;
    margin: 0 0.2rem;
    cursor: pointer;
  }
  &__indicator-separator {
    height: 0.8em;
    border-left: 1px solid #ddd;
    margin-left: 0.2em;
    padding-left: 0.2em;
  }
  &__loading-indicator {
    font-size: 1.4em;
    opacity: 0.3;
  }

  &__control {
    align-items: center;
    display: flex;
    justify-content: space-between;
    outline: 0 !important;
    position: relative;
    @extend %input;
    &--is-focused {
      outline: none;
      border: 1px solid #555 !important;
    }
  }

  &__value-container {
    display: flex;
    flex: 1;
    align-items: center;
    position: relative;
  }
  &__single-value {
    flex: 1;
    width: 0;
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
  }
  &__menu {
    position: absolute;
    background-color: hsl(0, 0%, 100%);
    box-shadow: 0 0 0 1px hsla(0, 0%, 0%, 0.1), 0 4px 11px hsla(0, 0%, 0%, 0.1);
    z-index: 1;
    width: 100%;
  }
  &__menu-list {
    max-height: 24rem;
    overflow-y: auto;
    position: relative;
  }
  &__option {
    &:not(:first-child) {
      border-top: 1px solid #ddd;
    }
    padding: 0.2rem 0.3rem;
    cursor: pointer;
    background: #fff;
    &--is-active {
      background-color: $active-color;
    }
    &--is-focused {
      background-color: $focused-color;
    }
    &--is-selected {
      color: white;
      background-color: $selected-color;
    }
  }
  &__group {
    border-top: 1px solid #888;
    padding: 0.5em 0;
  }
  &__group-heading {
    font-size: 0.8rem;
    text-transform: uppercase;
    color: #888;
    padding: 0.2rem 0.3rem;
    &:after {
      content: ' :';
      font-weight: bold;
    }
  }
  &__placeholder {
    opacity: 0.4;
  }
}
