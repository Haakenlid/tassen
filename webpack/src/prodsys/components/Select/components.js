import xSvg from 'images/x.svg'
import chevronSvg from 'images/chevron.svg'
import Throbber from 'components/Throbber'

export const DropdownIndicator = ({ innerProps }) => (
  <img className="react-select__indicator" src={chevronSvg} {...innerProps} />
)
export const ClearIndicator = ({ innerProps }) => (
  <img className="react-select__indicator" src={xSvg} {...innerProps} />
)
export const LoadingIndicator = ({ innerProps }) => (
  <span className="react-select__loading-indicator" {...innerProps}>
    <Throbber />
  </span>
)
