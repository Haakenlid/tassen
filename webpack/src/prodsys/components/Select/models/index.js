import * as contributors from './contributors.js'
import * as issues from './issues.js'
import * as storytypes from './storytypes.js'
import * as photos from './photos.js'
import * as stories from './stories.js'
import * as positions from './positions.js'

export default { positions, contributors, issues, storytypes, photos, stories }
