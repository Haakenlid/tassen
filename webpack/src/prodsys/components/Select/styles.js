const getStyles = () => {
  const styleKeys = [
    'container' /* main component */,
    // MAIN INPUT
    'control',
    'valueContainer',
    'input',
    'singleValue',
    'placeholder',
    // MENU
    // 'menu',  /* keep for automatic flip */
    'menuList',
    'option',
    'group',
    'groupHeading',
    // INDICATORS
    'indicatorsContainer',
    'indicatorSeparator',
    'dropdownIndicator',
    'loadingIndicator',
    'clearIndicator',
    // MULTIVALUE
    // 'multiValue',
    // 'multiValueLabel',
    // 'multiValueRemove',
    // MESSAGES
    // 'loadingMessage',
    // 'noOptionsMessage',
  ]

  const styles = {}
  for (const key of styleKeys) styles[key] = () => ({})
  styles['menu'] = R.pick(['top', 'bottom'])
  return styles
}

export default getStyles()
