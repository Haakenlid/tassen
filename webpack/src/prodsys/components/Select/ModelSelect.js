import Select from './Select.js'
import { connect } from 'react-redux'
import { modelSelectors, modelActions } from 'ducks/basemodel.js'
import models from './models'

const specFilter = R.ifElse(
  R.either(R.not, R.isEmpty),
  R.always(R.identity),
  R.pipe(R.toPairs, R.map(R.apply(R.propEq)), R.allPass, R.filter),
)

const mapStateToProps = (state, { model, filter, value }) => {
  const { getItems, getFetching } = modelSelectors(model)
  const { reshape = R.identity, reshapeOptions = R.identity } =
    models[model] || {}

  const items = R.map(reshape, getItems(state))
  const options = R.pipe(R.values, specFilter(filter), reshapeOptions)(items)
  const fetching = getFetching(state)
  const item = R.propOr(null, value, items)
  return { item, items, options, fetching }
}
const mapDispatchToProps = (dispatch, { model }) => {
  const { itemsRequested, itemRequested, itemCreated } = modelActions(model)
  return {
    search: params => dispatch(itemsRequested(params, true)),
    create: data => dispatch(itemCreated(data)),
    fetch: id => dispatch(itemRequested(id)),
  }
}

const ModelSelect = connect(mapStateToProps, mapDispatchToProps)(Select)

export default ModelSelect
