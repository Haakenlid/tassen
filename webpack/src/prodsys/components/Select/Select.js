import ReactSelect from 'react-select'
import CreatableSelect from 'react-select/lib/Creatable'
import models from './models'
import styles from './styles.js'
import * as components from './components.js'

class Select extends React.Component {
  constructor(props) {
    super(props)
    const model = models[props.model] || {}
    this.styles = styles
    this.components = {
      ...components,
      ...(model.components || {}),
    }
    this.getOptionValue = ({ value, id }) => value || id
    this.getOptionLabel = ({ label, ...props }) =>
      label || R.join(', ', R.keys(props))
    this.onInputChange = this.onInputChange.bind(this)
    this.state = { isLoading: props.fetching }
    this.onCreateOption = value => {
      const postData = (model.create || R.objOf('value'))(value)
      this.props.create(postData)
      this.props.onChange(value)
    }
  }

  onInputChange(value, { action }) {
    const { debounce = 500, search, filter = {} } = this.props
    this.setState({ inputValue: undefined })
    if (!search) return
    clearTimeout(this.fetchDebounce)
    if (action != 'input-change' || value.length < 3) {
      this.setState({ isLoading: false })
    } else {
      this.setState({ isLoading: true })
      this.fetchDebounce = setTimeout(() => {
        search({ search: value, ...filter })
      }, debounce)
    }
  }

  componentDidUpdate(prevProps) {
    const { fetching, value, item, model, items } = this.props
    if (prevProps.fetching != fetching) this.setState({ isLoading: fetching })
    if (value && !item) {
      const createdItem = R.last(
        R.filter(R.propEq('label', value), R.values(items)),
      )
      if (createdItem) {
        this.props.onChange(this.getOptionValue(createdItem))
      }
    }
  }

  componentDidMount() {
    const { value, item, fetch } = this.props
    if (!fetch) return
    if (value && !item) fetch(value)
  }

  render() {
    const {
      creatable,
      noneditable,
      options,
      item,
      onChange,
      ...props
    } = this.props
    if (noneditable) {
      // for use with ModelField
      return item ? this.getOptionLabel(item) : 'â€“'
    }
    const SelectComponent = creatable ? CreatableSelect : ReactSelect
    return (
      <SelectComponent
        onInputChange={this.onInputChange}
        getOptionLabel={this.getOptionLabel}
        getOptionValue={this.getOptionValue}
        components={this.components}
        options={options}
        styles={this.styles}
        menuIsOpen={undefined}
        isLoading={this.state.isLoading}
        isClearable={true}
        classNamePrefix="react-select"
        placeholder="Velg ..."
        noOptionsMessage={() => 'Ingen treff'}
        loadingMessage={() => 'Laster inn...'}
        onChange={value => onChange(value && this.getOptionValue(value))}
        minMenuHeight={500}
        menuPlacement="auto"
        {...props}
        {...creatableProps}
        onCreateOption={this.onCreateOption}
        value={item}
        className="ReactSelect"
        inputValue={this.state.inputValue}
      />
    )
  }
}

const creatableProps = {
  /* Allow options to be created while the `isLoading` prop is true. Useful to
     prevent the "create new ..." option being displayed while async results are
     still being loaded. */
  allowCreateWhileLoading: false,
  /* Gets the label for the "create new ..." option in the menu. Is given the
     current input value. */
  // formatCreateLabel: string => Node,
  /* Determines whether the "create new ..." option should be displayed based on
     the current input value, select value and options array. */
  // isValidNewOption: (string, ValueType, OptionsType) => boolean,
  isValidNewOption: (string, value, options) => string.length > 4,
  /* Returns the data for the new option when it is created. Used to display the
     value, and is passed to `onChange`. */
  // getNewOptionData: (string, Node) => OptionType,
  getNewOptionData: (value, node) => {
    return { value, label: value }
  },
  /* If provided, this will be called with the input value when a new option is
     created, and `onChange` will **not** be called. Use this when you need more
     control over what happens when new options are created. */
  // onCreateOption?: string => void,
  /* Sets the position of the createOption element in your options list. Defaults to 'last' */
  // createOptionPosition: 'first' | 'last',
  createOptionPosition: 'first',
}

export default Select
