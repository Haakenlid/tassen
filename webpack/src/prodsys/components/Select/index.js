import { connect } from 'react-redux'
import { modelSelectors, modelActions } from 'ducks/basemodel.js'
import Select from './Select.js'
import './Select.scss'

export { Select } // unconnected

const mapStateToProps = (state, { model }) => ({
  items: modelSelectors(model).getItems(state),
})
const mapDispatchToProps = (dispatch, { model }) => ({
  search: params => dispatch(modelActions(model).itemsRequested(params)),
  fetch: id => dispatch(modelActions(model).itemRequested(id)),
})

export default connect(mapStateToProps, mapDispatchToProps)(Select)
