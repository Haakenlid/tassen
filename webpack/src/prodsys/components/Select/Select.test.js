import Select from './index'
import renderer from 'react-test-renderer'

describe('Select', () => {
  test('renders correctly', () => {
    const tree = renderer.create(<Select />).toJSON()
    expect(tree).toEqual(expect.any(Object))
    expect(tree).toMatchSnapshot()
  })
})
