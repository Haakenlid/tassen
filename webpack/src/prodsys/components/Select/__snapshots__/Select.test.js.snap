// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders correctly 1`] = `
<div
  className="css-0 ReactSelect"
  onKeyDown={[Function]}
>
  <div
    className="css-0 react-select__control"
    onMouseDown={[Function]}
    onTouchEnd={[Function]}
  >
    <div
      className="css-0 react-select__value-container"
    >
      <div
        className="css-0 react-select__placeholder"
      >
        Velg ...
      </div>
      <div
        className="css-0"
      >
        <div
          className="react-select__input"
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <input
            aria-autocomplete="list"
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect="off"
            disabled={false}
            id="react-select-2-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            spellCheck="false"
            style={
              Object {
                "background": 0,
                "border": 0,
                "boxSizing": "content-box",
                "color": "inherit",
                "fontSize": "inherit",
                "label": "input",
                "opacity": 1,
                "outline": 0,
                "padding": 0,
                "width": "1px",
              }
            }
            tabIndex="0"
            type="text"
            value=""
          />
          <div
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "scroll",
                "position": "absolute",
                "top": 0,
                "visibility": "hidden",
                "whiteSpace": "pre",
              }
            }
          >
            
          </div>
        </div>
      </div>
    </div>
    <div
      className="css-0 react-select__indicators"
    >
      <span
        className="css-0 react-select__indicator-separator"
      />
      <img
        aria-hidden="true"
        className="react-select__indicator"
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
        src="test-file-stub"
      />
    </div>
  </div>
</div>
`;
